using_restricted_array.rb

```diff
@@ -7,41 +7,122 @@
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
 def length(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  until array[i] == nil
+    array[i]
+    i += 1
+  end
+  return i
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  until array[i] == nil
+    puts array[i]
+    i += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  if length == 1
+    return array[0] == value_to_find
+  else
+    i = 0
+    while i < length
+      if array[i] == value_to_find
+        return true
+      else
+        i += 1
+      end
+    end
+    return false
+  end
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  if length == 0
+    return nil
+  elsif length == 1
+    return array[0]
+  else
+    i = 0
+    greatest = array[0]
+    while i < length
+      if array[i] > greatest
+        greatest = array[i]
+      else
+        i += 1
+      end
+    end
+    return greatest
+  end
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  if length == 0
+    return nil
+  elsif length == 1
+    return array[0]
+  else
+    i = 0
+    smallest = array[0]
+    while i < length
+      if array[i] < smallest
+        smallest = array[i]
+      else
+        i += 1
+      end
+    end
+    return smallest
+  end
 end
 
 # Reverses the values in the integer array in place
 def reverse(array, length)
-  puts "NOT IMPLEMENTED"
+  if length == 0
+    return nil
+  elsif length == 1
+    return array[0]
+  else
+    i = 0
+    j = length - 1
+    while i < j
+      array[i] = array[i] + array[j]
+      array[j] = array[i] - array[j]
+      array[i] = array[i] - array[j]
+      i += 1
+      j -= 1
+    end
+  end
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def binary_search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  if length == 1
```

Comment:
Please take a look at the updated binary_search slide. 
Here's the other consideration I was missing in the algorithm on the slide before the update: Consider the scenario where there are two elements in the array and the value_to_find is the second element in the array. In the first iteration, min = 0, max=1, mid=0. Value is not at index 0. So, min = mid (still 0). On re-entering the loop, mid would be still 0. So we had an infinite loop. (To fix this, update max to be mid -1, if value_to_find is less than mid. And update min = mid+1, if value_to_find is greater than mid.)